File: ruby/lib/path_sum_backtracking.rb
class PathSumBacktracking
  
  def initialize(root, target)
    @root = root
    @target = target
    @result = []
  end

  def self.call(root, target)
    new(root, target).call
  end

  def call
    backtrack(@root, [], 0)
    @result
  end

  def backtrack(node, path, total)
    return unless node

    path << node.val
    total += node.val

    if total > @target
      path.pop
      return
    end
    
    if node.leaf? && total == @target
      @result = path.dup
    else
      backtrack(node.left, path, total)
      backtrack(node.right, path, total)
    end

    path.pop
  end
end

File: typescript/src/pathSumBacktracking.ts
import { Node } from './node'

export class PathSumBacktracking {
    root: Node
    number: number
    path: number[] = []

    constructor(root: Node, number: number) {
        this.root = root
        this.number = number
    }

    call(): boolean {
        return this.backtrack(this.root, 0)
    }

    backtrack(node: Node | undefined, acc: number): boolean {
        if (!node) return false

        this.path.push(node.val)
        acc += node.val

        if (acc > this.number) {
            this.path.pop()
            return false
        }

        if (node.isLeaf() && acc === this.number) return true

        const found = (
            this.backtrack(node.left, acc) ||
            this.backtrack(node.right, acc)
        )
        if (found) return found

        this.path.pop()
        return false
    }
}